### ISSUE TITLE
Balancing the contents of multiple columns to the same vertical space

### ISSUE THREAD

**Issue by ghost** [typst/typst#466]
I remember that in LaTeX the output of multicolumns will be balanced to the same vertical space in all columns to have enough space below that could start a new single column layout or an image or so.

In Typst the output for

```typst
#set par(justify: true)

#set page(columns: 2)

#lorem(500)
```

is currently unbalanced which looks not so professional, I think.

![balanced](https://user-images.githubusercontent.com/107212151/228925442-2115e896-ba41-4689-a8f1-e72f88f4d4aa.PNG)

Maybe a parameter like `balance-columns` or `columns: (balance: bool)` can be added to make this possible.

**Comment by myrrlyn** [comment]
I am interested in giving this a shot. I *think* I have a decent first draft of the requisite behavior. I'm going to describe it here to see if I'm on the right track:

My understanding (which I *hope* is correct!) is that a finitely-sized `Region` is a single PDF page, and that each `Frame` printed into a `Region` represents one rendered column of content. Assuming this is true, then my path forward is:

1. guard the first `body.layout()` call with `vt.provider.save()`/`vt.provider.restore()`. This gives me a collection of `Frame`s which I will then measure, and undoes any state changes that might have occurred during layout.
2. Discard all the `Frames` in that collection which are not in the final `Region`. This is just `frames.into_iter().skip(total_regions.saturating_sub(1) * columns)`.
3. Find the average rendered height of the remaining `Frames`. This is now the goal height of the final `Region`.
4. Reconstruct a laying-out backlog, using the full `regions.size.y` height for the `Frame`s in all but the final `Region`, and then switching over to the goal height for all subsequent `Frame`s.
5. Layout `body` into *that* backlog to produce a *new* collection of `Frame`s. This is the part that worries me: this layouting will need to be willing to overshoot the goal height, because it *cannot* be allowed to undershoot so much that additional `Frame`s are produced.

I *think* this should result in a collection of laid-out `Frame`s that are (almost) correctly balanced, and then the `Region` assembly loop can be left untouched.

However, my attempt at writing this out does not result in column-balancing behavior in the test suite. I've only just started reading the codebase, so I'm not at all informed as to where I should be looking to understand why the expected effects aren't occurring. I'll continue to investigate this on my own and see if I can figure out where my assumptions are failing to meet reality.

**Comment by xosxos** [comment]
Bump.. it would be very nice to have single column images and two column text mixed together

**Comment by astrale-sharp** [comment]
Or a `tight: true` option

**Comment by laurmaedje** [comment]
I would call it `balanced`. `#set columns(balanced: true)`.

**Comment by eduardz1** [comment]
any developments?

**Comment by KDr2** [comment]
any developments?

**Comment by laurmaedje** [comment]
Not yet

**Comment by Caellian** [comment]
In the meantime, a solution that works for content that doesn't span multiple pages is:

```typst
#let balanced-cols(content) = style(styles => {
  let h = measure(content, styles).height / 2
  block(height: h, columns(2, content))
})
```

**Comment by astrale-sharp** [comment]
`#balanced-cols(lorem(200))`
![image](https://github.com/typst/typst/assets/53686698/d8c4d34c-13d6-4079-8823-acb76ed1dcca)

Color me unimpressed :rofl: 

more srsly, Am I using it wrong?

**Comment by Caellian** [comment]
Yeah, inline content doesn't have a fixed height. AFAICT `measure`d height is min required height for content so you end up with `h == 0` because a paragraph doesn't have a min height.

This is what happens when I add newlines:
![image](https://github.com/typst/typst/assets/4082795/bb09abb8-02f5-415a-9368-1ca67e92d6a4)

I wrapped it in blocks because using boxes (inline element) breaks in a similar way. `balanced-cols` only works well when you have uniform children blocks that fit on a single page. I can't see any viable solution for text though.

**Comment by dahooz** [comment]
A possible workaround could be:

```typst
#let eqcolumns(n, gutter: 4%, content) = {
  layout(size => [
    #let (height,) = measure(
      block(
        width: (1/n) * size.width * (1 - float(gutter)*n), 
        content
      )
    )
    #block(
      height: height / n,
      columns(n, gutter: gutter, content)
    )
  ])
}
```

This uses `#layout` to get the surrounding width (`size.width`), calculates the column width depending on `gutter` and column number `n` and typesets the content under that width in a `#block`. Then it takes the `#measure`d height of this block, and typesets the actual `#columns` in a `#block` of the calculated `height` divided by the column number `n`, so that the `#columns` has a forced height, which forces it to reflow the content accordingly.

This works for one-paragraph text and reasonably well for multi-paragraph-text (it looks ugly if a paragraph break is close to  a column break, because it does not treat the column break as replacing the paragraph break). It will also break if the total content height is longer than one page!

**Comment by soleilvermeil** [comment]
+1
For the simple reason that one of typst's mottos is "Compose papers faster", and noticing that most articles are written in two columns, I find it surprising that this feature isn't already implemented.

@dahooz's workaround is interesting, but does not work great for papers of over a page, which is unfortunate.

**Comment by Caellian** [comment]
Looking at the `FlowLayouter` impls, there seems to be enough information there to implement `eqcolumns` in a way that works across pages.

I'll give a go to implementing this in a week or so, my progress is [here](https://github.com/Caellian/typst).
I have to compliment the devs, the codebase is absolutely beautiful at the first glance (can't wait to ruin it :laughing:).

**Comment by xosxos** [comment]
@Caellian If you haven't noticed, a refactor of `flow.rs` was just merged yesterday https://github.com/typst/typst/pull/4931 and I couldn't rebase your code. Just a heads up in case your current ideas for the draft would be going against the grain.

**Comment by laurmaedje** [comment]
Yes, the flow layout is indeed changing and will change a lot more soon. It might be best to wait until after that.

**Comment by Caellian** [comment]
> Yes, the flow layout is indeed changing and will change a lot more soon. It might be best to wait until after that.

I'll hold off until then. My current implementation separates balancing into a wrapper struct that calls flow layout. Should I integrate it directly into flow layout instead if used private functions get spread across multiple modules?

Ideally, I'd like to be able to measure vertical height of each element given the width constraint of the Regions they're located in, and it's a bit tricky because that requires them to have a Location which they don't until flow layout runs, and I can foresee cases where that Location would become invalidated by balancing. If possible it would be very useful for that part of flow layout to be able to run separately for this purpose and possibly some other in the future.

**Comment by laurmaedje** [comment]
> Should I integrate it directly into flow layout instead if used private functions get spread across multiple modules?

Let's continue that discussion once the changes have landed! I think it'll be much more actionable then.

**Comment by laurmaedje** [comment]
@Caellian The flow rewrite is merged now (#5017), so feel free to take a look at that.

The main feature is that layout of a single region can be restarted now, which I think could be quite useful for balancing columns.

The new preprocessed children representation is also much easier to deal with and measure because it pre-assigns Locations.

**Comment by laurmaedje** [comment]
> I hope I won't need to do that unless I encounter some weird edge cases. Maybe for pages with columns. But the #columns element should in theory be able to layout children evenly in a single pass given that regions stay the same across pages and element height depends on region width which makes them both constant.

Could you elaborate on your planned approach?

**Comment by Caellian** [comment]
> > I hope I won't need to do that unless I encounter some weird edge cases. Maybe for pages with columns. But the #columns element should in theory be able to layout children evenly in a single pass given that regions stay the same across pages and element height depends on region width which makes them both constant.
> 
> Could you elaborate on your planned approach?

I was planning on measuring children [here](https://github.com/Caellian/typst/blob/249037c097aaf256cd7637ed7f94a05155c0b3e5/crates/typst/src/layout/flow/mod.rs#L214), and then limiting how they get distributed by externally modifying `Work.children` slice [here](https://github.com/Caellian/typst/blob/249037c097aaf256cd7637ed7f94a05155c0b3e5/crates/typst/src/layout/flow/mod.rs#L223) via a struct that has [special drop handling](https://github.com/Caellian/typst/blob/249037c097aaf256cd7637ed7f94a05155c0b3e5/crates/typst/src/layout/flow/balance.rs#L71). But I just figured out that styles can be modified from inside columns which would make those measurements wrong - so I'll need to layout those children at least twice.

**Comment by Wulfheart** [comment]
Seems quite complicated. I would expect something like columns(balance: true).

**Comment by laurmaedje** [comment]
> Seems quite complicated. I would expect something like columns(balance: true).

The discussion is about how it's implemented, not how it looks like for a user.

> I was planning on measuring children [here](https://github.com/Caellian/typst/blob/249037c097aaf256cd7637ed7f94a05155c0b3e5/crates/typst/src/layout/flow/mod.rs#L214), and then limiting how they get distributed by externally modifying Work.children slice [here](https://github.com/Caellian/typst/blob/249037c097aaf256cd7637ed7f94a05155c0b3e5/crates/typst/src/layout/flow/mod.rs#L223) via a struct that has [special drop handling](https://github.com/Caellian/typst/blob/249037c097aaf256cd7637ed7f94a05155c0b3e5/crates/typst/src/layout/flow/balance.rs#L71). But I just figured out that styles can be modified from inside columns which would make those measurements wrong - so I'll need to layout those children at least twice.

The nice thing about relayouting a region is that the measurements you get from layouting within the region are much more accurate and true to nature than if you'd try to measure upfront, especially with blocks breaking over pages.

**Comment by Wulfheart** [comment]
> The discussion is about how it's implemented, not how it looks like for a user.

Thank you for your hard work! Didn’t have the full context.

**Comment by iagobaapellaniz** [comment]
IMHO, columns should be balanced by default. One always can use `V(123pt)` to get any desired result.

**Comment by Wulfheart** [comment]
Or somehow unbalance them.

**Comment by Caellian** [comment]
> IMHO, columns should be balanced by default

I don't agree, changing this default would change most existing multicolumn documents. Also, balancing is an "extra thing" the way I view it - it's computationally more expensive and would (presumably) be have to turned off for majority of use cases because afaik only scientific articles are balanced. Counter examples include: menus in restaurants, pamphlets, books, etc.

**Comment by iagobaapellaniz** [comment]
I've never been in restaurants then

**Comment by iagobaapellaniz** [comment]
This seems to do the job for weak balancing columns
```
#let balance(content) = layout(size => {
  let count = content.at("count")
  let textheight = measure(content).at("height")
  let linegap = par.leading.em * textheight
  let (height,) = measure(block(width: size.width, content),)
  let lines = calc.ceil((height - textheight) / count / (textheight + linegap))
  let newheight = lines * (textheight + linegap) + textheight
  [#block(height: newheight)[#content]]
})

#balance(columns(n)[#lorem(80)])
```

You can se it here
https://typst.app/project/ryVROoNxih2oBQC7dNwc4V

**Comment by xosxos** [comment]
Very happy to see this moving forward! 

I agree on principle that if the balancing increases compile times, it should not be on by default. _However_, I also can't imagine a common case where you wouldn't want a balanced layout when _switching_ the number of columns. Books etc tend to be single column, and n=1 should not require any extra compute?

_Also_, I am assuming that adding a tiny single column paragraph to the end of a VERY LARGE two column document, does not require computing over the entire VERY LARGE document to balance for the tiny single column. Otherwise, this could lead to a surprising performance drop.

**Comment by Caellian** [comment]
> I've never been in restaurants then.

Ok, well... they often use multicolumn layout for dishes, where segments don't have breaks and given space restrictions/cost, the last column always has tailing break.

Your provided example has no vertical constraints so of course the columns will fill the first column, it does look bad, but if you set max height for them they will render fine (which would ideally be propagated from page).

> Also, I am assuming that adding a tiny single column paragraph to the end of a VERY LARGE two column document, does not require computing over the entire VERY LARGE document to balance for the tiny single column. Otherwise, this could lead to a surprising performance drop.

If the whole document is 2(+)-columned then it would need to be entirely re-layouted. If it's added to a child `#columns` then that `#columns` needs to be and any elements after it. Multicolumn balancing generally approaches the ideal layout in `O(log n)` relayouts - speed of each depends on discrete use case and element counts.

**Comment by iagobaapellaniz** [comment]
The restaurant comment was irony. Menus need to mix layouts, and without a constraint `columns` take all available vertical space.

Besides that, hyphenation also needs computation. Orphan paragraph too. Almost everything needs computation. Besides that, Typst has wonderful `text`, `par`, `block`, ... functions with a lot of styling features. `columns` is too basic (int, gutter, content) for what I would expect.

**Comment by dfyx** [comment]
> This seems to do the job for weakly balanced columns
> 
> #let balance(content) = layout(size => {
>   let count = content.at("count")
>   let textheight = measure(content).at("height")
>   let linegap = par.leading.em * textheight
>   let (height,) = measure(block(width: size.width, content),)
>   let lines = calc.ceil((height - textheight) / count / (textheight + linegap))
>   let newheight = lines * (textheight + linegap) + textheight
>   [#block(height: newheight)[#content]]
> })
> 
> #balance(columns(n)[#lorem(80)])
> You can se it here https://typst.app/project/ryVROoNxih2oBQC7dNwc4V

This works great for simple cases. Unfortunately, as soon as the content contains headings, it breaks.

**Comment by dayala1** [comment]
*THE* issue, in my opinion. Having balanced columns is what would allow to change columns on the go to achieve appealing layouts.

**Comment by pacaunt** [comment]
Current workaround that I came up with: 
```typst 
#let resolve-length(len, max: 0pt) = {
  if type(len) == length { return len }
  if type(len) == ratio { return max * len }
  if type(len) == fraction { return len / 1fr * max }
  panic("Cannot resolve the length " + repr(len))
}

#let eqcols(n, body, shift: 0cm, gutter: 4%, debug: false) = {
  let eqcols-id = counter("__eqcols")
  eqcols-id.step()
  context {
    let floating-figures = state("__figs_" + str(eqcols-id.get().at(0)), ())
    show figure.where(scope: "parent"): it => { floating-figures.update(x => x + (it,)) + it }

    layout(size => {
      let container-width = size.width
      let container-height = size.height
      let line-height = 1em.to-absolute()
      let gutter = resolve-length(gutter, max: container-width)
      let shift = resolve-length(shift, max: container-height)
      let column-width = { (container-width - (n - 1) * gutter) / n }

      let float-free-body = {
        show figure.where(scope: "parent"): none
        body
      }

      let float-free-one-column-size = measure(width: column-width, float-free-body)

      context {
        let all-floating-figures = floating-figures.final()
        let float-height = all-floating-figures
          .map(fig => measure(width: container-width, fig).height)
          .sum(default: 0pt)
        let raw-new-height = float-free-one-column-size.height / n + float-height + shift
        let new-height = calc.ceil((raw-new-height / 1pt) / (line-height / 1pt)) * line-height
        block(columns(n, body, gutter: gutter), height: new-height, stroke: if debug { red })
      }
    })
  }
}
```
This can handle floating figures and balance columns across pages. However, this cannot handle manual column breaks.

**Comment by git-staus** [comment]
Sorry for adding noise, but I just wanted to cheer for the recent activity on this issue. I started using Typst a few weeks ago and I got here via [this](https://forum.typst.app/t/1915) forum thread from several years ago (which topped my search results). I am trying to introduce it in a professional environment and this issue has been the largest eyesore in my two-page reports. I am very happy to see things like @eltos' PR progressing.

I just wanted to say that when the Typst maintainers find the time to take this further, at least one person's day would be made (mine!). Thank you all for all the good work you do!

**Comment by gamd** [comment]
> balancing is an "extra thing" the way I view it - it's computationally more expensive...

I understand that technology itself may shape new aesthetics, but since when do we want to regard for computer's feelings over ours? The need for columns in typesetting has been out there way longer than typst or computers, and they are almost always balanced.
